Refactor the settings pages, and add unit tests for them

Konamiman opened the pull request:
**Note:** [The original pull request](https://github.com/woocommerce/woocommerce/pull/26024) has been closed because it was very old and there was no way to sensibly merge the changes from master, not because of merge conflicts (there were just a handful and not very complex) but because of issues with the code sniffer (it was complaining about one of the configured sniffs being unknown, and complaining about each and every .js file in the codebase).

### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?


### Changes proposed in this Pull Request:

This pull request introduces an extensive refactor of the WooCommerce settings classes (`WC_Settings_Page` and derived classes) that includes the following:

* Modify the `get_settings()` method so that it accepts a section id as a parameter, but instead of including this parameter in the method signature get it with `func_get_arg`

  * Some derived classes in the WooCommerce code had changed the method signature to `get_settings( $current_section = '' )` already, but others (those not having multiple sections natively) hadn't, making then effectively impossible to define multiple sections for these pages via filters.
  * `func_get_arg` is needed for compatibility with PHP 8 since some external derived classes may have redefined the signature to `get_settings( $current_section = '' )` but others will still be using `get_settings()`, see the comment added to the method for details.
  * Because of the described hack `get_settings` is marked as deprecated, `get_settings_for_section` should be used instead.
<br/>

* Add a `get_settings_for_section($section_id)` method in the base class that searches for a `get_settings_for_{section_id}_section` method in the class and executes it, otherwise it executes the new protected method `get_settings_for_section_core( $current_section )`; then it triggers the `'woocommerce_get_settings_' . id` filter.

  * This makes it easier to separate the code that returns the list of filters in multiple methods, one per section, instead of using one big if-else-else... block.
  * Also, `'woocommerce_get_settings_' . id` is triggered in one single place too.
  *  `get_settings_for_section` is marked as `final` because we want derived classes to either implement `get_settings_for_{section_id}_section` methods or override the `get_settings_for_section_core` method (or both).
<br/>

* General refactorings to fix code sniffer errors, including replacing old-style sniffer directives with phpcs:enable/disable/ignore comments.

* Some duplicated code has been refactored to protected methods.

* Changes the "todo: remove in 4.0" comment in `WC_Settings_Rest_API` to "remove in 5.0".


### How to manually test the changes in this Pull Request:

Create the following plugin and open the _General_ settings tab. You'll see that it has three sections: `General, Foobar, Fizzbuz`. The first one has one extra setting, the second one has just one setting, and the thirs one is empty. Without the changes introduced in this pull request you would have been able to create the sections, but the `woocommerce_get_settings_general` trigger would be invoked with only the `$settings` parameter and thus you wouldn't have been able to return different per section settings.

```php
<?php
/*
Plugin Name: Foobar & Fizzbuzz
Version: 1.0
Description: Foo? Bar. Fizz? Buzz. See? It's easy!
Author: WooCommerce (testing)
*/

add_filter('woocommerce_get_sections_general', 'foobar_sections_for_general_settings', 10, 1);

function foobar_sections_for_general_settings($settings) {
	$settings['foobars'] = 'Foobars';
	$settings['fizzbuzzs'] = 'Fizzbuzzs';
	return $settings;
}

add_filter('woocommerce_get_settings_general', 'foobar_settings_for_general_tab', 10, 2);

function foobar_settings_for_general_tab($settings, $current_section) {
	if( '' === $current_section ) {
		$settings[] =
			array(
				'title' => "Bonus settings",
				'type'  => 'title',
				'desc'  => "Some bonus settings inside the default section!"
			);

		$settings[] =
			array(
				'title'    => 'Some setting',
				'desc'     => 'Is this setting set or not?',
				'id'       => 'foobar_some_setting',
				'default'  => 'no',
				'type'     => 'checkbox',
				'desc_tip' => 'A setting that can be set or not.'
			);

		$settings[] = array( 'type' => 'sectionend'	);

		return $settings;
	}
	else if( 'foobars' === $current_section ) {
		return array(
			array(
				'title' => "Foobars",
				'type'  => 'title',
				'desc'  => "What about foobars?"
			),

			array(
				'title'    => 'The foo is bar',
				'desc'     => 'States if the foo is bar',
				'id'       => 'foobar_the_foo_is_bar',
				'default'  => 'no',
				'type'     => 'checkbox',
				'desc_tip' => 'Is the foo bar enough?',
			),

			array( 'type' => 'sectionend' )
		);
	}
	else {
		return array();
	}
}
```

It's also important to test with a settings class that directly overrides the `get_settings` method, [here's a handy code snippet for that](https://gist.github.com/claudiosanches/618892ba1c27c48e8298dab374d71216), or you can use an existing extension like [Freshbooks](https://github.com/woocommerce/woocommerce-freshbooks/blob/917a57b8851becd7dc62b39e99a198e4dd527d39/includes/admin/class-wc-freshbooks-settings.php).

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?


### Changelog entry

> Enhancement - All settings pages can now be extended consistently with new sections and settings. Also, unit tests have been added.

Konamiman committed:
Refactor the settings pages, and add unit tests for them.

This commit fixes some inconsistencies in the settings pages, and
makes all the existing pages extensible by adding new sections
(that was possible in some pages, but not in others). Main changes:

1. Modify the 'get_sections' method so that it invokes a new protected
   'get_own_sections' method and then triggers the
   'woocommerce_get_sections_' . id filter.

This way the filter is triggered only in the base class
and not in each of the derived classes too.

2. Change the get_settings() method so that it has its signature
   changed to get_settings( $current_section = '' )
   in the base class and in all the derived class.

Some derived classes were already using this signature, but others
(those not having multiple sections natively) weren't, making then
effectively impossible to define multiple sections for these pages
via filters.

With this change all the section pages act consistently and allow
both adding new settings to the default "General" section
and creating new sections via filters.

3. Change the implementation of 'get_settings' in the base class
   so that it searches for a 'get_settings_for_{section_id}_section'
   method in the class and executes it, otherwise it executes the new
   protected method get_settings_for_section( $current_section ); then
   it triggers the 'woocommerce_get_settings_' . id filter.

This makes it easier to separate the code that returns the list
of filters in multiple methods, one per section, instead of using
one big if-else-else... block.

So now instead of overriding get_settings($current_section='') derived
classes need to implement get_settings_for_{$current_section}_section
for each section, or override get_settings_for_section($current_section)
or both. 'get_settings_for_section' returns an empty array by default.

Also, 'woocommerce_get_settings_' . id is triggered in one single
place too.

Other improvements:

* Remove duplicated code from 'output' in 'WC_Settings_Page' children.

Some classes inherited from 'WC_Settings_Page' override the 'output'
method with custom code, which in all cases ended up repeating the code
of the original method as a fallback. These repetitions have been
replaced with 'parent::output()'.

* Fix inconsistencies for 'save' and 'output' in WC_Settings_Tax/Emails

The 'WC_Settings_Tax' and 'WC_Settings_Emails' classes had some
inconsistencies in their 'save' and 'output' methods that prevented the
proper creation new sections and the addition of new settings via the
'woocommerce_get_sections_' and 'woocommerce_get_settings_' filters.
Now they work as expected.

* Deduplicate parts of 'save' in 'WC_Settings_Page' and children.

Two methods have been added to 'WC_Settings_Page' class:
'save_settings_for_current_section' and 'do_update_options_action'.
These are intended to be invoked by derived classes in their 'save'
methods, in order to remove code repetition.

* Add some helper methods to WC_Unit_Test_Case.

Methods added:
- assertOutputsHTML
- assertEqualsHTML
- normalize_html
- capture_output_from

rodrigoprimo commented:
@Konamiman before someone review this PR could you please fix Travis and the conflicting files?

Konamiman commented:
Hi @rodrigoprimo, the PR branch is now up to date with master, thanks.

Konamiman committed:
Fix unit test problems in PHP 7.0.

This includes removing two array_merge and array_merge_eecursive
statements in the code hacker, that apparently were working differently
in PHP 7.0.

Konamiman commented:
@rodrigoprimo I had to make some extra changes for tests to pass in PHP 7.0, they all pass now.

claudiosanches reviewed (changes requested):
There's some conflicts to fix.

Konamiman committed:
Fix unit tests after merge from master

Konamiman commented:
Hi @claudiosanches, I've fixed the conflicts but now there are some weird unit test failures. It's tricky because the failures only appear when running all the tests, but not when running the failing file only. I took a look but all I can see is that some orders are created with a post type of just `post`, instead of `shop_order`... do you have any clue of what can be going on?

Konamiman committed:
Move code hacker resetting from BeforeTestHook to setUp

The code hacker needs to be reset before each test. This was done via
a couple of classes implementeing BeforeTestHook, those were registered
in phpunit.xml.

The problem is that the PHPUnit version used for WooCommerce unit test
has recently been changed from 7.5 to 6.5 for compatibility with
PHP 7.0, and hook classes were introduced in PHPUnit 7. Thus no hooks
were ran, the code hacker wasn't reset, that caused some functions
to remain hacked between tests, and this made some tests to fail.

The solution is to move the code hacker reset to the setUp method
in the base unit test class.

Konamiman committed:
Re-implement the flushing of the term count cache that was lost on merge

Also add unit tests for it.

Konamiman commented:
Hi @claudiosanches, everything is fixed and ready for review now.

claudiosanches reviewed (changes requested):
Seems like it caused some bugs in the shipping options page, located in `/wp-admin/admin.php?page=wc-settings&tab=shipping&section=options`

> PHP Fatal error:  Uncaught Error: Call to undefined method WC_Settings_Shipping::output_fields() in wp-content/plugins/woocommerce/includes/admin/settings/class-wc-settings-shipping.php:185

Konamiman committed:
Fix wrong handling of the shipping - shipping options section.

Konamiman commented:
Hi @claudiosanches, good catch, thanks! I believe this was probably caused by some badly resolved merge conflict. It's fixed now,.

Konamiman commented:
Hi @claudiosanches, I've fixed yet another conflict that had popped, could you review again plz?

Konamiman committed:
Fix unit tests after merging from trunk

claudiosanches commented on includes/admin/settings/class-wc-settings-advanced.php:
```diff
@@ -31,375 +31,410 @@ public function __construct() {
 			'webhooks'        => __( 'Webhooks', 'woocommerce' ),
 			'legacy_api'      => __( 'Legacy API', 'woocommerce' ),
 			'woocommerce_com' => __( 'WooCommerce.com', 'woocommerce' ),
+			'features'        => __( 'Features', 'woocommerce' ),
```
There's no `features` menu by default, this one now introduces a blank page.

claudiosanches committed:
Remove "Features" menu

claudiosanches committed:
Remove todo

claudiosanches reviewed (dismissed):
Great job on this one, I have tested in many ways I could, I'm going to list some of my testing process and considerations, but I think we could merge this one.

In one of my tests I used this documentation: https://docs.woocommerce.com/document/adding-a-section-to-a-settings-tab/
While in this docs talk about "Products" tab, I changed it to the general tab, where there's no sections, and it started to fail in "trunk", but works now just fine in this branch.
You can test with: https://gist.github.com/claudiosanches/77bb7fd0b4ed44f22775289cea189189
So great job for making this work in all tabs!

Also seems that our REST API didn't got impacted by any issue, I tested all endpoints from `wp-json/wc/v3/settings/` and seems to be working fine.

And now the only issue with this PR is that could throw PHP warnings in PHP 7.4 and a PHP Fatal error in PHP 8.0 if someone is extending `WC_Settings_Page` and inserting a `get_settings()` method without the `$current_section` parameter.

Those are the errors that you can get:

```
[PHP 7.4] PHP Warning:  Declaration of class@anonymous::get_settings() should be compatible with WC_Settings_Page::get_settings($current_section = '') in wp-content/plugins/test-plugin/test-plugin.php on line 64
[PHP 8.0] PHP Fatal error:  Declaration of WC_Settings_Page@anonymous::get_settings() must be compatible with WC_Settings_Page::get_settings($current_section = '') in wp-content/plugins/test-plugin/test-plugin.php on line 64
```

Here the code I used to test: https://gist.github.com/claudiosanches/618892ba1c27c48e8298dab374d71216

I also tested with some extensions from our marketplace, and we could devise those in 3 categories:

1. Extensions that integrated by filters ready to maintain compatibility with anything, like: [Store credits](https://github.com/woocommerce/woocommerce-store-credit/blob/6a1303a962c2ce39c690436161f7f54b579801cf/includes/admin/class-wc-store-credit-admin-settings.php)
2. Extensions that are just luck enough to use sections: [Subscriptions Advanced Options](https://github.com/woocommerce/woocommerce-subscriptions-advanced-options/blob/22ac7b85c320e41acefaac28daa7e7ca1e28b2f2/includes/admin/class-wcsao-admin-settings.php)
3. And most common case, extensions that are following what we did in the past and just extending `WC_Settings_Page::get_settings()` without any parameter, like: [Freshbooks](https://github.com/woocommerce/woocommerce-freshbooks/blob/917a57b8851becd7dc62b39e99a198e4dd527d39/includes/admin/class-wc-freshbooks-settings.php).

If you like I can list other extensions that I tested, but by implementing this PR we'll have to update and patch some of our extensions.

claudiosanches committed:
Fixed unit tests

Konamiman commented:
@claudiosanches - very good catch. This is a compatibility issue we can't overlook.

I think the proper solution for this would be:

1. Rename the existing base method `get_settings( $current_section = '' )` to `get_settings_for( $current_section = '' )`. This is also consistent with the protected `get_settings_for_section` and `get_settings_for_X_section` that I introduced in the PR.
2. Modify the existing `get_settings` as follows (and also maybe deprecate it):

```php
public function get_settings() {
    return $this->get_settings_for('');
}
```

This seems to work well in PHP 8 in all cases: derived settings classes implementing `get_settings()` and those implementing `get_settings( $current_section = '' )`.

wdyt? Is there any potential problem with this approach?

claudiosanches commented:
@Konamiman I think is a good idea, and could works just fine, and I agree that we could include a `@deprecated` flag in `get_settings()`.

Konamiman committed:
Adjustments in settings pages after a merge with conflicts.

Konamiman committed:
Rename "get_settings" to "get_settings_for_section" in settings pages.

In PHP 8 overriding a method having an optional parameter with a
method having no parameters throws an error, thus we can't use
the strategy of changing "get_settings()" to "get_settings($section='')"
without breaking existing extensions. So we do the following instead:

- Rename the existing "get_settings" to "get_settings_for_section"
- Rename the existing "get_settings_for_section" to "get_settings_for_section_core"
- Add a "get_settings" that just does "get_settings_for_section('')"
  for compatibility, but mark it as deprecated.

Konamiman committed:
Make the WC_Settings_Page::get_settings_for_section final

This helps on conveying the notion that the method to be overriden
is get_settings_for_section_core instead (or get_settings_for_X_section
methods must be added).

Konamiman committed:
Refactor in the WC_Settings_Page class for PHP 8 compatibility

- Turn get_settings into a parameterless method, but accept one
  parameter via func_get_arg; and mark the method as deprecated.
- Rename the existing get_settings to get_settings_for_section;
  and mark the method as final.
- Rename the existing get_settings_for_section to get_settings_for_section_core.

See the comment added to get_settings for the rationale for the change.

ObliviousHarmony commented on includes/admin/settings/class-wc-settings-shipping.php:
```diff
@@ -64,81 +64,105 @@ public function get_sections() {
+	 *
+	 * @return bool True is the WC_INSTALLING constant is defined.
+	 */
+	protected function wc_is_installing() {
```
Technically this isn't actually necessary. The Jetpack Constants package we're using lets you change constants for unit tests already. This may not have been the case when you wrote the PR I think.

Konamiman committed:
Merge branch 'trunk' into refactor/settings-pages-classes-take-2

Konamiman committed:
Fix syntax error after manual merge conglict resolution

claudiosanches reviewed (changes requested):
Hello @Konamiman,

This PR is backwards incompatible, now isn't possible to just extend `WC_Settings_Page` and create a method called `get_settings()` to load all your settings as before.

I found out it by testing again all my steps from:
https://github.com/woocommerce/woocommerce/pull/27684#pullrequestreview-611071775


You can reproduce it by testing with my snippet: https://gist.github.com/claudiosanches/618892ba1c27c48e8298dab374d71216

Or by installing any of our extensions that extends `WC_Settings_Page` class, like [Freshbooks](https://github.com/woocommerce/woocommerce-freshbooks/blob/917a57b8851becd7dc62b39e99a198e4dd527d39/includes/admin/class-wc-freshbooks-settings.php)

Konamiman committed:
Fix: no output for settings classes that directly override get_settings.

Konamiman commented:
@claudiosanches very good catch, many thanks! Fortunately that one was easy to fix. I've also added a couple of additional unit tests covering this case, and I've updated the PR testing instructions with a link to your snippet and the Freshbooks extension.

claudiosanches reviewed (changes requested):
Everything seems to be working as expected now, I couldn't find any bug this time.
But I agree with Christopher: https://github.com/woocommerce/woocommerce/pull/27684/files#r627491960

claudiosanches reviewed (approved):
Seems finally ready to go!
